src/XrdOss/XrdOssApi.cc

```diff
@@ -254,6 +259,77 @@ int XrdOssSys::GenRemotePath(const char *oldp, char *newp)
     return 0;
 }
 
+/******************************************************************************/
+/*                                  C l o n e                                 */
+/******************************************************************************/
+/*
+  Function: Clone contents of a file from another file.
+
+  Input:    srcFile     - clone contents of this file.
+
+  Output:   Returns XrdOssOK upon success and -errno or -osserr upon failure.
+*/
+
+int XrdOssFile::Clone(XrdOssDF& srcFile)
+{
+#if defined(FICLONE)
+  if (!canClone)
+    return -EPERM;
+
+  if (fd<0)
+     return -EBADF;
+  if (srcFile.getFD() < 0)
+     return -EBADF;
+
+  if (ioctl(fd, FICLONE, srcFile.getFD())==-1)
+    return -errno;
+
+  return XrdOssOK;
+#else
+  return -ENOTSUP;
+#endif
+}
+
+int XrdOssFile::Clone(XrdOucCloneSeg cVec[], int n)
+{
+#if defined(FICLONERANGE)
+  if (!canClone)
+    return -EPERM;
+
+  if (n < 0)
+     return -EINVAL;
+
+  if (fd<0)
+     return -EBADF;
+
+  if (n == 0)
+    return XrdOssOK;
+
+   int i, j=0;
+  do{
+   int sFd = cVec[j].srcFD;
+
+   for (i = j+1; i < n && cVec[i].srcFD == sFd; i++);
```

Comment:
Given that the ioctl() can only do one element at a time, I don't think you need anything more that a simple loop that goes through the coloc vector and issues the ioctl() for each element. So, this can be dramatically simplified by using a single simple iterative loop.